sdk/python/kfp/registry/client.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2022 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class for KFP Registry Client."""
+
+import logging
+
+import google.auth
+import json
+import requests
+import re
+from google.auth.credentials import Credentials
+from typing import Any, Optional, List, Tuple, Union
+
+_KNOWN_HOSTS_REGEX = {
+    "kfp_pkg_dev":
+        r'(^https\:\/\/(?P<location>[\w\-]+)\-kfp\.pkg\.dev\/(?P<project_id>.*)\/(?P<repo_id>.*))',
+}
+
+_DEFAULT_JSON_HEADER = {
+    'Content-type': 'application/json',
+}
+
+
+class _SafeDict(dict):
+
+    def __missing__(self, key: str):
+        return '{' + key + '}'
+
+
+class ApiAuth(requests.auth.AuthBase):
+
+    def __init__(self, token: str):
+        self._token = token
+
+    def __call__(self, request: requests.Request):
+        request.headers['authorization'] = 'Bearer ' + self._token
+        return request
+
+
+class RegistryClient:
+
+    def __init__(self,
+                 host: str,
+                 auth: Optional[Union[requests.auth.AuthBase,
+                                      Credentials]] = None):
+        self._host = host.rstrip('/')
+        self._known_host_key = ""
+        for key in _KNOWN_HOSTS_REGEX.keys():
+            if re.match(_KNOWN_HOSTS_REGEX[key], self._host):
+                self._known_host_key = key
+                break
+        self._config = self.load_config()
+        if auth:
+            self._auth = auth
+        elif self._is_ar_host():
+            self._auth, _ = google.auth.default()
+
+    def _request(self,
+                 request_url: str,
+                 request_body: str = '',
+                 http_request: str = 'get',
+                 extra_headers: dict = None) -> Any:
+        """Call the HTTP request"""
+        self._refresh_creds()
+        auth = self._get_auth()
+        http_request_fn = getattr(requests, http_request)
+
+        response = http_request_fn(
+            url=request_url,
+            data=request_body,
+            headers=extra_headers,
+            auth=auth)
+        response.raise_for_status()
+
+        return response
+
+    def _is_ar_host(self):
+        return self._known_host_key == "kfp_pkg_dev"
+
+    def _is_known_host(self) -> bool:
+        return bool(self._known_host_key)
+
+    def load_config(self):
+        config = {}
+        if self._is_ar_host():
```

Comment:
can you extract any platform specific code from this file to a dedicated file?

Reply:
As per the description of the PR, platform specific code here is temporary (i.e. functionality for supporting config files will be done as a follow-up PR, and this will be extracted to a separate config file). Added a todo.